arangod/Aql/AstResources.cpp

```diff
@@ -44,15 +44,38 @@ AstResources::AstResources(arangodb::ResourceMonitor& resourceMonitor)
       _shortStringStorage(_resourceMonitor, 1024) {}
 
 AstResources::~AstResources() {
-  // free strings
+  clear();
+  size_t memoryUsage = _strings.capacity() * memoryUsageForStringBlock();
+  _resourceMonitor.decreaseMemoryUsage(memoryUsage);
+}
+
+// frees all data
+void AstResources::clear() noexcept {
+  size_t memoryUsage = (_nodes.numUsed() * sizeof(AstNode)) + _stringsLength;
+  _nodes.clear();
+  clearStrings();
+  _resourceMonitor.decreaseMemoryUsage(memoryUsage);
+
+  _shortStringStorage.clear();
+}
+
+// frees most data (keeps a bit of memory around to avoid later re-allocations)
+void AstResources::clearMost() noexcept {
+  size_t memoryUsage = (_nodes.numUsed() * sizeof(AstNode)) + _stringsLength;
+  _nodes.clearMost();
+  clearStrings();
+  _resourceMonitor.decreaseMemoryUsage(memoryUsage);
+
+  _shortStringStorage.clearMost();
+}
+
+// clears dynamic string data
+void AstResources::clearStrings() noexcept {
   for (auto& it : _strings) {
```

Comment:
I like the fact that there is now a `clearStrings`. However, memory usage must be updated by **the called** of `clearStrings`, this is unintuitive and error-prone. I know that we do not want to add another atomic operation, so we should at least add a clearly visible comment to `clearStrings` that the **caller** has to adjust the memory usage **before** calling `clearStrings` by `_stringsLength`. Both current call sites do this, so there is no bug, but this is dangereous,.

Reply:
Ok, will do!

Reply:
Made `clearStrings()` private and added code comments.